File: src/client/powermemo-ts/src/network.ts
import { ServerError } from './error';
import type { BaseResponse } from './types';

export async function unpackResponse<T>(response: Response): Promise<BaseResponse<T>> {
  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`);
  }

  const data = (await response.json()) as BaseResponse<T>;

  if (data.errno !== 0) {
    throw new ServerError(data.errmsg);
  }

  return data;
}

File: src/client/powermemo-go/network/network.go
package network

import (
	"encoding/json"
	"net/http"
	
	mberror "github.com/memodb-io/powermemo/src/client/powermemo-go/error"
)

type BaseResponse struct {
	Data   map[string]interface{} `json:"data"`
	Errmsg string                 `json:"errmsg"`
	Errno  int                    `json:"errno"`
}

func UnpackResponse(resp *http.Response) (*BaseResponse, error) {
	if resp.StatusCode >= 400 {
		return nil, &mberror.ServerError{Message: resp.Status}
	}

	var baseResp BaseResponse
	if err := json.NewDecoder(resp.Body).Decode(&baseResp); err != nil {
		return nil, err
	}

	if baseResp.Errno != 0 {
		return nil, &mberror.ServerError{Message: baseResp.Errmsg}
	}

	return &baseResp, nil
} 